### PR TITLE
PoC: memory tracker v2

### PR DESCRIPTION
**Pull request by BohuTANG**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Track the memory usage of one tokio runtime.

Fixes #issue

### CODE REVIEW

**Comment by xudong963** [comment]
Does the ticket aim to track rumtime memory usage?

**Comment by BohuTANG** [comment]
> Does the ticket aim to track rumtime memory usage?

Yes, try to do that.

File: src/common/base/src/runtime/allocator.rs [hunk 1]
```diff
@@ -0,0 +1,411 @@
...
+                        SKIP_CACHE[i] |= cur_bit;
+                        true
+                    } else {
+                        CHECKED_CACHE[i] |= cur_bit;
+                        false
+                    }
+                }
+            });
+            if verbose {
+                tracing::info!(?stack, "STARTED_TRACE");
```
Comment (Xuanwo) [inline src/common/base/src/runtime/allocator.rs:370]: This call will call alloc again, we should avoid call tracing in allocator.

Comment (BohuTANG) [reply src/common/base/src/runtime/allocator.rs:370]: PoC phase, will remove them all when it's ready.

Comment (Xuanwo) [reply src/common/base/src/runtime/allocator.rs:370]: > PoC phase, will remove them all when it's ready.

I'm guessing we will get `SIGABRT` here.

File: src/common/base/src/mem_allocator/allocator.rs [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        Self { inner }
+    }
+}
+
+unsafe impl<A: GlobalAlloc> GlobalAlloc for ProxyAllocator<A> {
+    #[inline]
+    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
+        let tid = get_tid();
+        let slot = tid % COUNTERS_SIZE;
+        MEM_SIZE[slot].fetch_add(layout.size(), Ordering::Relaxed);
```
Comment (zhang2014) [inline src/common/base/src/mem_allocator/allocator.rs:80]: is very slow.

File: src/common/base/src/mem_allocator/allocator.rs [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        MEM_CNT[slot].fetch_add(1, Ordering::Relaxed);
+
+        self.inner.alloc(layout)
+    }
+
+    #[inline]
+    unsafe fn dealloc(&self, ptr: *mut u8, layout: Layout) {
+        let tid = get_tid();
+        let slot = tid % COUNTERS_SIZE;
+        MEM_SIZE[slot].fetch_sub(layout.size(), Ordering::Relaxed);
```
Comment (zhang2014) [inline src/common/base/src/mem_allocator/allocator.rs:90]: crash will occur when the destroyed memory is larger than the allocated memory.

**Comment by BohuTANG** [comment]
@zhang2014 This is a PoC last week, it seems failed, I will close this PR.

**Pull request closed without merging**